src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs

```diff
@@ -0,0 +1,478 @@
+            }
+
+            TaskCompletionSource tcsConnect = new TaskCompletionSource();
+
+            // For Abort/Dispose.  Calling Abort on the request at any point will close the connection.
+            _cts.Token.Register(s => ((BrowserWebSocket)s!).AbortRequest(), this);
+
+            try
+            {
+                if (requestedSubProtocols?.Count > 0)
+                {
+                    using (JavaScript.Array subProtocols = new JavaScript.Array())
+                    {
+                        foreach (string item in requestedSubProtocols)
+                        {
+                            subProtocols.Push(item);
+                        }
+                        _innerWebSocket = new HostObject("WebSocket", uri.ToString(), subProtocols);
+                    }
+                }
+                else
+                {
+                    _innerWebSocket = new HostObject("WebSocket", uri.ToString());
+                }
+                _innerWebSocket.SetObjectProperty("binaryType", "arraybuffer");
+
+                // Setup the onError callback
+                _onError = new Action<JSObject>((errorEvt) =>
+                {
+                    errorEvt.Dispose();
+                });
+
+                // Attach the onError callback
+                _innerWebSocket.SetObjectProperty("onerror", _onError);
+
+                // Setup the onClose callback
+                _onClose = new Action<JSObject>((closeEvt) =>
+                {
+                    using (closeEvt)
+                    {
+                        _innerWebSocketCloseStatus = (WebSocketCloseStatus)closeEvt.GetObjectProperty("code");
+                        _innerWebSocketCloseStatusDescription = closeEvt.GetObjectProperty("reason")?.ToString();
+                        _receiveMessageQueue.Writer.TryWrite(new ReceivePayload(Array.Empty<byte>(), WebSocketMessageType.Close));
+                        NativeCleanup();
+                        _tcsClose?.SetResult();
+                    }
+                });
+
+                // Attach the onClose callback
+                _innerWebSocket.SetObjectProperty("onclose", _onClose);
+
+                // Setup the onOpen callback
+                _onOpen = new Action<JSObject>((evt) =>
+                {
+                    using (evt)
+                    {
+                        if (!cancellationToken.IsCancellationRequested)
+                        {
+                            // Change internal _state to 'Connected' to enable the other methods
+                            if (Interlocked.CompareExchange(ref _state, (int)InternalState.Connected, (int)InternalState.Connecting) != (int)InternalState.Connecting)
+                            {
+                                // Aborted/Disposed during connect.
+                                throw new ObjectDisposedException(GetType().FullName);
+                            }
+                            tcsConnect.SetResult();
```

Comment:
`onopen` will not be called if the JS `WebSocket` fails the handshake, instead only the `onerror` and `onclose` event will be called. Which means that the `tcsConnect` will never complete in this case.
Should `onerror` or `onclose` also try to set an exception on `tcsConnect`?

Example:
```js
var ws = new WebSocket("wss://github.com");
ws.onopen = e => console.log("open", e);
ws.onerror = e => console.log("error", e);
ws.onclose = e => console.log("close", e);
```
```
error Event {isTrusted: true, type: "error", target: WebSocket, currentTarget: WebSocket, eventPhase: 2, …}
close CloseEvent {isTrusted: true, wasClean: false, code: 1006, reason: "", type: "close", …}
```

Reply:
Yes, that is the case.  Have not noticed that before.

Reply:
Ah, I see now.  The code that set the tcsConnect result was removed from close callback.  I need to put it back.  Thanks for the catch.

Reply:
This is now fixed.